Feature: Water tiles have a depth

nielsmh opened the pull request:
Allocate 4 bits in M3 to store the depth of MP_WATER tiles.

In the base game, depth affects the cost of clearing (and otherwise modifying) water tiles, and the meaning of "river" versus "sea" speed for ships has changed to "shallow" versus "deep" water.
There is support for NewGRFs to supply a table of water surface sprites for different depth levels, via action 5. Maybe this should be added to the baseset.
Additionally, NewGRF industries will be able to check the water depth, so e.g. dredging sites in FIRS could be limited to construction on shallow water.

Water depth 0 and 1 have the original cost to clear, higher depths have the cost to clear multiplied by the square of depth, so clearing the max depth of 15 (represented in land area information window as 300 m) has a cost of around £1.7 million at default 'low' construction costs, while clearing a water tile at the coast costs £7,524.
Rivers are soft limited to a depth of 3 (60 m), and canals are always built at depth 0 (10 m).

Water tiles can "erode" over time to even out the depth gradient from the shore. This is an option that can be disabled, or the frequency can be selected.

## Not in scope
- Depth-aware bridges. Doing an overhaul of bridges is a large project in itself and is probably best tackled in a later patch.
- Multiple variations of water tiles. Having more than one tile graphics for each depth level could reduce visual repetition, but is a large project in itself.
- Adding deep water tiles to the original baseset, although it should probably be done before a release is made with this patch included.

## To do
- [x] Canals built over rivers should probably inherit the depth of the river.
- [x] Evening out of depth and depth gradient should be more controlled and slower -- maybe a game setting to control the speed or entirely disable it.
- [x] Need a way for GS and AI to query depth.
- [x] Need a way for NewGRF to query depth, especially for industries constructing on water.
- [x] Perhaps change the ship "on river/canal" speed limit to count as "on shallow water" instead.
- [x] Would be neat to have different water tiles for different depths (perhaps darker blue).
- [x] Configurable cost of modifying deep water tiles.
- [ ] Need a test case of rivers/canals GRF using new variable for depth.
- [x] Find a way to make ground sprites below oil rig be depth aware.

## Unresolved questions
- Should industry tiles and station tiles also have space to store water depth? Right now, industry tiles do not store water depth, instead the industry itself stores the min and max water depths it was on when it was constructed. Stations (docks, buoys, oil rig stations) store nothing and instead search for a nearby water tile whose depth is assumed.
- What should the default setting for `difficulty.water_clearing_cost_exponent` be? Is the "basic" category appropriate?
- What should the default setting for `difficulty.water_depth_erosion_speed` be? Is the "advanced" category appropriate?
- What values should those settings be given when savegames are converted?

## NewGRF docs that need to be updated with this PR
- Action 5 adds feature 0x19 to define deep water terrain sprites. Requires an array of sprite offsets as "other data" mapping depth levels to sprites.
- Canal/river feature callbacks add variable 0x84 to query the water depth of the current tile.
- Industry placement callback has two new error returns available, 0x409 and 0x40A for respectively "water too deep" and "water too shallow".
- Industry callbacks add two new variables, 0xAD and 0xAE, to get the minimum and maximum water depth on the site the industry was/is going to be constructed.
- Ships feature callbacks add variable 0xE6 to get water depth of current tile.

nielsmh committed:
Feature: Water tiles have a depth

nielsmh committed:
Add: Water depth evens out over time

nielsmh committed:
Change: Use square of water depth for clear cost multiplier

nielsmh committed:
Change: Restore water depth from surrounding tiles when clearing buoys, oil rigs, etc.

James103 commented:
I have a couple of points to make regarding customizing this feature:
1. Could you add a way to disable, lessen, (or even increase) the cost multiplier of clearing water tiles?
2. Could you add a way to disable the water depth check when building ship depots?
3. Could you add a way to reduce the max water depth from level 15 (even all the way to 0 to disable this feature)?

nielsmh commented:
@James103 Point 1 and 2, maybe. For cost of clearing water, the setting could be between "no multiplier", "more expensive by depth" and "much more expensive by depth" with the last being my current square-of-depth multiplier.

Point 3 I don't think is a good idea. NewGRF support will be more complex if depth levels can be arbitrarily limited.

LordAro commented:
[![](https://imgs.xkcd.com/comics/workflow.png)](https://xkcd.com/1172/)

We cannot add options to disable every single change to the game.

nielsmh commented:
This, along with the earlier "neutral stations only serve their owning industry" change are support for changing the long-standing TT tradition of filling in the ocean. With deep ocean being expensive to fill in, building bridges crossing large water bodies as short segments with intermediate islands for signals, is also much less viable, making transfer to ships or longer routes around the water body more attractive. In all, I think this change will make for much more interesting gameplay.

sheepo99 commented:
> This, along with the earlier "neutral stations only serve their owning industry" change are support for changing the long-standing TT tradition of filling in the ocean. With deep ocean being expensive to fill in, building bridges crossing large water bodies as short segments with intermediate islands for signals, is also much less viable, making transfer to ships or longer routes around the water body more attractive. In all, I think this change will make for much more interesting gameplay.

I sense this would also imply the need for more vessel types for navigating diff depths and balancing ships overall, meaning new vehicle graphics of some sort would be necessary as well. It would also make canals for bigger ships much more expensive, privileging small-ship networks for inland canals. So many interesting options!

James103 commented:
With this change, clearing a large enough body of water will cause the total cost to roll past 64 bits, resulting in an overflow and possibly gaining money.
* Cost to clear water tile of depth 0: 7,500
* Cost multiplier from depth 15: x225
* Maximum map size: x16,777,216 tiles
* Max NewGRF cost multiplier: x65,536
* Max inflation multiplier: x887
* Total Cost to clear All-Water 4k map: 1,855,425,871,872,000,000
* Total Cost to clear All-Water 4k map after inflation: 1,647,156,981,589,051,266,457
* Maximum Tiles before 64-bit Overflow (after inflation): 93,945 tiles
* MAX_INT64 (signed):  9,223,372,036,854,775,807

Therefore: there **is a risk** of a 64-bit overflow after this change, but it only applies when estimating the cost in the Scenario Editor. If the cost multiplier per unit of depth were just 0.5-1 higher than what it is now (depth^2.5 or depth^3), then said risk would become **real** as the number of tiles required to be demolished to trigger a 64-bit overflow would now fit within the maximum tile clearing limits.

nielsmh commented:
I'm glad to hear that the cost of filling in the ocean is finally getting reasonable. I don't think the above shows a problem with this PR, rather if an underflow can occur that's a bug in the Money class. Also, wouldn't you generally hit the landscaping limit throttle before the underflow?

nielsmh committed:
Change: Make ship pathfinder consider shallow water as rivers, and deep rivers as sea

nielsmh committed:
Change: Canals built over rivers retain depth

nielsmh committed:
Change: Tile loop water depth propagation 64 times slower

SamuXarick commented on src/water_cmd.cpp:
```diff
@@ -247,6 +257,14 @@ void MakeWaterKeepingClass(TileIndex tile, Owner o)
+	}
+	if (min_water_depth <= WATER_DEPTH_MAX) SetWaterDepth(tile, min_water_depth);
+
 	if (wc != WATER_CLASS_INVALID) CheckForDockingTile(tile);
```
```
	if (wc != WATER_CLASS_INVALID) {
		/* Restore the water depth from surrounding tiles */
		uint8 min_water_depth = WATER_DEPTH_MAX + 1;
		for (Direction dir = DIR_BEGIN; dir < DIR_END; dir++) {
			const TileIndex dest = tile + TileOffsByDir(dir);
			if (IsValidTile(dest) && IsTileType(dest, MP_WATER)) min_water_depth = min(min_water_depth, GetWaterDepth(dest));
		}
		if (min_water_depth <= WATER_DEPTH_MAX) SetWaterDepth(tile, min_water_depth);

		CheckForDockingTile(tile);
	}
```

SamuXarick replied:
Industries being removed near coasts can trigger an assert in SetWaterDepth, due to the surroundings being a coast with depth 0.

Edit: this is the suggested fix.

nielsmh committed:
Change: Use effective water depth instead of effective water class to determine ship speeds

nielsmh committed:
Codechange: Move depth erosion code to a separate function

nielsmh committed:
Change: Do full water depth erosion during map gen

nielsmh committed:
Change: Use full map erosion to add water depth to old saves

nielsmh committed:
Fix: Crash when land industries next to shore tiles are cleared

nielsmh committed:
Add: TGP makes deep water on its own

nielsmh committed:
Add: [Debug] show water depth in zoomed in views in debug builds

James103 commented:
Looks like the regression tests are failing (as of 8f59276). Can you please check where the regression tests are failing and why they are failing?

SamuXarick commented:
they fail because it cannot place a ship depot

nielsmh commented:
What @SamuXarick  says. I'm looking into a bug with how water depth is restored when a ship depot gets cleared, the problem is that it doesn't. (So you can glitch depth by filling a deep area with depots, clearing them, and then filling in the sea to avoid a few millions in cost.)

nielsmh committed:
Fix: Regressions broken by ship depot depth check, and keep depth correctly when removing ship depots

James103 commented:
The regression tests are still failing. Why? Is it because of not enough money? Additionally, the macOS regression test timed out after 1 hour. The logs say that the regression test AI "script died unexpectedly".

nielsmh commented:
The regression is failing because the program crashes. The program crashes because of an assertion error. The assertion error is that a tile that was expected to be water was not water. The assertion happens during destruction of a ship depot.

The macOS build hangs because it does not handle crashes in the regressions tests.

floodious commented:
> With this change, clearing a large enough body of water will cause the total cost to roll past 64 bits, resulting in an overflow and possibly gaining money.

There are multiple ways to handle such a situation. One is to brute-force compute the sum by using a bignum accumulator that grows by adding additional words as needed. This requires an additional check for wrapping on each operation, once per additional N-bit word. Such bignum implementations already exist and their costs are known.

Alternatively it's possible to ensure within the known context that wrapping can never occur. Noting that only nonlinear operations like square/log/exp are noncommutative while multiplication is; if the sum is minimized by computing the sum of the squares of the height multiplier before computing the product:

(depth 0 == x1, therefore 15 == x16 ?)
Maximum cost multiplier from depth 16: 16^2 == 256 (2^8)
Maximum map size: 4096^2 == 16777216 tiles (2^24)
Maximum multiplicand/sum: 256 x 16777216 = 4294967296 (== 2^32)

This sum can then be used to compute the product via the bignum code (128 bit or otherwise) including the coefficient for Cost x Inflation. This means the overhead of the bignum product would only apply to a single step at the end of the summation stage and the sum itself would fit well within a 64 bit accumulator.

nielsmh commented:
OTTD already uses a Money type which is supposed to be safe against overflow (by saturating at INT64_MAX), but does not check for underflow. That needs to be fixed in [src/core/overflowsafe_type.hpp](https://github.com/OpenTTD/OpenTTD/blob/master/src/core/overflowsafe_type.hpp) and is not a discussion for this PR.

floodious commented:
Right, that was going to be my first suggestion to use saturating arithmetic, but that can't possibly compute the correct sum and it's generally less efficient without hardware support. So I thought, the only way to actually compute the accurate sum+product is to use a multi-step process where the sum and product are handled separately with adequate precision for each.

floodious commented:
Regarding the actual cost scaling:

![squares](https://user-images.githubusercontent.com/59541655/72340274-bfcdc680-367c-11ea-971a-0090f52f1f23.png)

Assuming the surrounding vertices are equal height, at depth:
1. = 1 
2. = 3^2 + 1 == 10
3. = 5^2 + 3^2 + 1 == 35
4. = 7^2 + 5^2 + 3^2 + 1 == 84

To get the accurate scaling I can't see any way other than actually iterating through every vertex and summing the difference in height. x^2 seems ... I'm blanking out on the terms here, but what was the rationale for x-squared?

(Blanking out: the difference isn't merely polynomial but rather ... ? Making application of a polynomial "severely underestimate the growth of the cost with increasing magnitude of x." = word I'm looking for.)

Sure took me long enough to bother to remember: Factorial, hyper-(exponentiation, ...), tetration, pentation, ...

nielsmh commented:
I don't think having a fully correct arithmetic for pathological cases that are only possible on gigantic maps with inflation and landscaping costs maxed out is important. As long as it results in the player being refused the action because they would go into negative it's good enough. Even in the extreme case of the player having INT64_MAX cash on hand, and performing an action that should cost more than INT64_MAX, the action succeeding and the playing ending up at zero is also acceptable.

nielsmh commented:
My reason for using quadratic growth on sea clearing cost is that it's easy to implement and does a lot to make filling in the sea unattractive, not that it models anything resembling the real world.

floodious commented:
I'd need to look at whether the existing landscaping cost computation uses correct iteration and computation of height differences given the "1 per tile" rule. Computing the accurate cost if it isn't yet implemented may very well "solve" the need for limitations on landscaping by making them bear more realistic cost and scaling. On the other hand, it may also make more than very minor tweaks very cost prohibitive as they become way higher order than a polynomial.

edit to add:
I suppose the volume of a rectangular pyramid is the proper volume. In the square case the area would indeed be quadratic with the product of height in addition to that if I'm not mistaken, shouldn't the accurate maximum value then be `height^2*height/3` ? If so my thinking `x^^n` was wrong since it's only `(n*2+1)^2+((n-1)*2+1)^2+... == n^2*n/3`, apparently.
Then:
1. = 1/3, 2. = 8/3, 3. = 27/3, 4. = 64/3
So n^2 would be ... (sub- ... that word I'm looking for?) but only very slightly so and close to accurate.

edit again: (or isn't that then supposed to be n^3/3?)
Still can't recall this one: hypographic (underestimating) vs. epigraphic (overestimating) (?).

nielsmh committed:
Add: [NewGRF] Canal feature var 0x84 gets water depth

RoqueDeicide commented:
This change feels wrong to me, unless it comes with an option to make clearing of water tiles cost the same amount regardless of depth. This would allow marriage of new mechanics, like limiting ability to build depots, with classic gameplay without turning the game into one of those "hard-mode" patchpacks (I don't play on the latter for a reason).

nielsmh committed:
Add: [NewGRF] Action 5 feature 0x19 sets water sprites with depth

nielsmh committed:
Fix: [NewGRF] Various range bugs in depth water sprites support

nielsmh commented:
Here's two NewGRFs with deep water sprites, drawn by TT-Forums users wallyweb and Andrew350, and the NFO files adjusted by me.
[water-depth-newgrfs.zip](https://github.com/OpenTTD/OpenTTD/files/4067084/water-depth-newgrfs.zip)

The WaterCycleTest GRF is 8bpp and based on original baseset, uses palette animation. The darkwater GRF is 32bpp and based on OpenGFX, does not have palette animation but covers all 16 depth levels.

nielsmh committed:
Add: Setting to control cost of deep water modification

nielsmh committed:
Add: Script function to query water depth

nielsmh committed:
Change: Text for ship speed to match new behaviour

nielsmh committed:
Add: [NewGRF] Industry location check callback variable for water depth

nielsmh commented:
Added a NewGRF variable to get the water depth for industry callbacks, although I'm not sure this is the best way to do it. The variable will probably not be any use after the industry's construction since the tiles are now all industry instead of water, so it should perhaps only be available during the location check callback, in which case it could override one of the variable numbers currently occupied by various cargo statuses that are useless during location check.

nielsmh committed:
Remove: Industry var 0x72 again, it's a bad way of doing water depth for industries

nielsmh committed:
Add: [NewGRF] Industry var 0xAD and 0xAE for min and max water depth the industry was constructed over

nielsmh commented:
Should there perhaps be a game setting for water depth smoothing speed? No smoothing, slow, fast? Should smoothing perhaps be limited to a certain range of depths?

floodious commented:
Oceanic silt settling velocity analysis:
https://agupubs.onlinelibrary.wiley.com/doi/full/10.1002/jgrc.20086
> Accuracy of these results was generally better than for the Stokes' law, the Stokes' law modified with the Schiller‐Naumann drag coefficient, and Rubey's equation. The hypothesis that the settling velocity is a function of linear and nonlinear drag, and is ultimately determined by physical characteristics as much as biological composition and internal aggregate geometry, is supported here by quantitative analyses.

My best summary so far is that they're describing an equation to compute the terminal velocity for an aggregate of particles according to measured parameters rather than empirical parameters. The result seems to be coefficients used in a transcendental function with a non-linear feedback loop ... suffice it to say, we do have "good" approximations for this but the complexity and computational cost is high relative to a linear time-invariant solution. The effects you'd expect to see with regard to non-linear settling on the ocean floor can be modeled by a fairly simple function to produce realistic results; a linear time-invariant solution such as a constant settling rate will not produce such realism.

There are often middle-ground approaches that work well enough. The LTI variation would produce very poor results cheaply, while the full non-linear model would be prohibitively computationally expensive for full realism. A partially simplified model may provide the best trade-off including "realistic enough" non-linear results along with side-effects due to discretization. I believe computational cost is likely to be the primary concern.

floodious commented:
My suggestion would be to use an [exponential moving average filter](https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average) with a cheap low order non-linear function used to approximate the coefficient each tick. Essentially you have particles at the surface of the slope that get "stirred up" into a low-density solution. There is also a constant high-density terminal velocity which would eventually lead to complete settling of all particles.

The factors involved include "what material?" and "what density?" For example a solid stone foundation will settle very slowly (or not at all) while silt will settle quickly due to differences in particle size, friction, weight and density. Solid materials are the source of permanent "bumps" on the ocean floor with a silt layer on top that always settles to the point where settling force drops below friction, yielding zero velocity.

So we have at least four variables: solid height, (solid + ) silt height, silt density (state) and silt friction (material). The settling would occur via an iterative process applied to height + density according to a non-linear function: f(height delta, density delta, friction constant)

It may be possible to compute the effect as a 3x3 bitmap matrix filter with the coefficient matrix updated each time. This is similar to old style [water effects](http://agilerepose.weebly.com/water-ripple.html).

floodious commented:
Would it make sense to assume solid height is always equal to zero, except where river meets ocean and solid height == silt height? In such a case creeks and rivers would constantly "settle" into nearly level areas and form an ugly simplified type of deltas without a flow velocity model that creates the bends in rivers. The flat/straight deltas and beaches would settle into surrounding ocean and you should get a naturally emergent shelf near land.

In which case "ocean depth" is in fact silt depth... although that doesn't allow for any permanent variation in the solid depth. The coefficients would need to be set up to ensure that a slope delta == 1 doesn't constantly settle, otherwise you end with completely flat bottoms at depth == 1 driven by settling from the coasts/edges. That's why I suggested the additional variables of silt depth + density, where higher density/material would drive settling velocity toward zero without incoming low-density materials to fill the void.

Honestly the whole thing is questionable as the settling rate of these materials is very low and this is more a sort of geological time issue rather than a century from 1950 to 2050.

nielsmh commented:
A 3x3 filter is effectively what I use to perform the smoothing right now, what might be better would be a 5x5 or even 7x7 filter to allow lower gradients than 1 depth unit per tile. Also yes, making the erosion much slower. I looked at some nautical charts of the seas around Denmark and really most of it is below 60 metres of depth, very large parts even below 10. Only in Skagerrak between Norway and Denmark do depths go past 100 metres, as much as 700 metres.

My conclusion from this is, for regional-ish maps as I think most OpenTTD games use, a depth range of 10-300 metres is more than sufficient, and for maps covering entire continents or worlds, you may as well ignore depth and just use a basecosts-mod if you want to discourage filling in the ocean.

floodious commented:
With only four bits of precision I was thinking the main computation could occur applied to two additional bitmap arrays via a threaded process. This could be processed in small chunks with sleeps to try to control the rate... say updating the settling once per month or slower.

There could be an array containing integer or float variables for height + density and the output could be written to the 4-bit depth value only when a change has occurred during the update period. That would allow far higher precision, non-linearity and very gradual smooth changes with only the consequence of additional memory requirements.

With an iterative IIR filter given enough precision the matrix size doesn't actually influence the minimal gradient, as changes "spread out" each iteration. So like the water effect demo I linked it may be possible to take four samples rather than nine, or even operate on a by-row + by-column alternating basis taking two samples for each output.

For depth ranges I agree, a majority of the coast in the real world is approximately 100 to 200 meters deep along the continental shelf. During the last ice age the sea level would have been low enough to expose these regions around ~20k years ago. So what we think of as the coast today is certainly not a static thing at all over not just geological time but even thousands of years. For periods near a century the non-linearity means the largest effects will occur most quickly when changes are made to the coast. This is often seen as rapid erosion near construction or fills along coastal regions that happen rapidly on a year-to-year or by-decade basis.

nielsmh committed:
Doc: AI and GS changelog

nielsmh committed:
Add: Setting to control sea floor erosion speed

nielsmh committed:
Fix: Mark tile loop eroded tiles dirty

nielsmh committed:
Fix: Only erode tiles shallower by one level at a time

nielsmh commented:
Scrapping the idea of transition sprites between depth levels. There's too many combinations possible (240 with just a single edge considering a single neighbour) for it to be realistic to make art for, and it would create excessively complex NewGRF callbacks too. Making it table-based would also be ridiculous.

Instead, added a to-do item about water sprites below oil rigs.
![image](https://user-images.githubusercontent.com/1062071/79669886-2d7b1880-81bf-11ea-86ae-ece7575809b4.png)

nielsmh committed:
Change: Allow GetWaterDepth to also work on industry tiles

nielsmh commented:
Would it be crazy to stuff water depth into industry tiles m6, split like bit 6-7 and 1-0?
It could also be added to m3 for station tiles, at least the water station tiles.

nielsmh committed:
Change: Also use GetWaterDepth for industry tiles when a ship queries

nielsmh committed:
Add: Ship callbacks variable 0xE6 to get current water depth

nielsmh committed:
Change: Cache depth of tile ship is on for speed calculation and NewGRF purposes

This is to avoid risks of desync if the depth changes while a ship is on a tile, and one client
considers the old depth and another client (which arrived just now) considers the new depth.

nielsmh commented:
Here's a small GRF (made with a quickly hacked up NML) that tests the new properties for industry construction and ships. It seems to work.
[deepwatertest.zip](https://github.com/OpenTTD/OpenTTD/files/4505431/deepwatertest.zip)

nielsmh commented:
Perhaps I should also add a `GSTile::SetWaterDepth` function to let GS modify the depth, but that will either require a new command or have it piggyback on an existing landscaping command.

SamuXarick commented on src/saveload/afterload.cpp:
```diff
@@ -3100,6 +3100,28 @@ bool AfterLoadGame()
+			const IndustrySpec *spec = GetIndustrySpec(ind->type);
+			ind->water_depth_min = ind->water_depth_max = 0;
+			if (spec == nullptr || !(spec->behaviour & INDUSTRYBEH_BUILT_ONWATER)) continue;
+			/* This industry type builds on water, try to find nearby water tiles for synthetic depth */
+			TileIndex tile = INVALID_TILE;
+			if (CircularTileSearch(&tile, 5, [](TileIndex t, void *) { return IsWaterTile(t); }, nullptr)) {
+				ind->water_depth_min = ind->water_depth_max = GetWaterDepth(tile);
+			}
```
IsWaterTile requires valid tiles, but you pass INVALID_TILE to it, and it asserts.

I think something like this might work:
```
			/* This industry type builds on water, try to find nearby water tiles for synthetic depth */
			TileIndex tile = INVALID_TILE;
			TILE_AREA_LOOP(tile_cur, ind->location) {
				tile = tile_cur;
				if (CircularTileSearch(&tile, 5, [](TileIndex t, void*) { return IsWaterTile(t); }, nullptr)) {
					ind->water_depth_min = ind->water_depth_max = GetWaterDepth(tile);
				}
			}
```

PeterN commented:
Seems to be out of date ;(

perezdidac commented:
https://www.enr.com/articles/44013-the-10-longest-floating-bridges-in-the-world I don't want to be too chatty, but how about floating bridges 👯

nielsmh commented:
Rebased so it builds again, and added the preview label. Right now the preview builds won't have any graphics for deep water, so and the fallback of printing a depth number on each tile is only used in debug builds. You can view the depth with the query tool.
I see there's a couple integer sign and width issues to fix.
